### PR TITLE
Unify wasm compilation flags

### PR DESCRIPTION
**Pull request by tiziano88**
This removes some of the logic around compiling Wasm modules to a
specific out dir, but I think it makes more sense to keep the default
location and modify the app manifest files to point to it rather than
convincing cargo to put compiled artifacts in non-standard folders, even
if the manifests end up being slightly longer.

- Fix the path of some of the internal dependencies, since some tests
started randomly failing, presumably because of some interference with
externally published crates with the same names (namely `prost`).

### CODE REVIEW

File: sdk/rust/oak_tests/src/lib.rs [hunk 1]
```diff
@@ -100,61 +106,43 @@ const RETRY_INTERVAL: std::time::Duration = std::time::Duration::from_millis(800
 /// Convenience helper to build and run a single-Node Application with the given module name, using
```
Comment (rbehjati) [inline sdk/rust/oak_tests/src/lib.rs:106]: The comment still mentions the module name. Can you update it?

File: sdk/rust/oak_app_build/src/main.rs [hunk 2]
```diff
@@ -299,6 +299,9 @@ async fn main() -> anyhow::Result<()> {
 
     let output_file = get_output_path(&manifest_dir, &manifest.name);
 
+    std::fs::create_dir_all(manifest_dir.join(OUTPUT_DIRECTORY))
```
Comment (rbehjati) [inline sdk/rust/oak_app_build/src/main.rs:302]: Could this be moved into `get_output_path`? Since `get_output_path` is only called here.

File: sdk/rust/oak_tests/src/lib.rs [hunk 3]
```diff
@@ -38,33 +38,42 @@ pub enum Profile {
...
+        package_name
+    ));
+    module_path
+        .to_str()
+        .expect("could not convert PathBuf to String")
+        .to_string()
+}
 
+/// Uses cargo to compile a Rust manifest to Wasm bytes.
+pub fn compile_rust_wasm(manifest_path: &str, profile: Profile) -> anyhow::Result<Vec<u8>> {
```
Comment (rbehjati) [inline sdk/rust/oak_tests/src/lib.rs:65]: Does the function with the same name in `oak_functions` need to be updated too? Perhaps in a separate PR, or as part of #1965.

File: sdk/rust/oak_tests/src/lib.rs [hunk 4]
```diff
@@ -100,61 +106,43 @@ const RETRY_INTERVAL: std::time::Duration = std::time::Duration::from_millis(800
...
-    run_single_module_with_config(
-        module_wasm_file_name,
-        entrypoint_name,
-        ConfigMap::default(),
-        permissions,
-    )
+    run_single_module_with_config(entrypoint_name, ConfigMap::default(), permissions)
 }
 
 /// Convenience helper to build and run a single-Node application with the given Wasm module file
```
Comment (rbehjati) [inline sdk/rust/oak_tests/src/lib.rs:123]: Update comment?

File: sdk/rust/oak_tests/src/lib.rs [hunk 5]
```diff
@@ -100,61 +106,43 @@ const RETRY_INTERVAL: std::time::Duration = std::time::Duration::from_millis(800
...
-        entrypoint_name,
-        config_map,
-        permissions,
-    );
+    let combined_config = runtime_config(entrypoint_name, config_map, permissions);
     oak_runtime::configure_and_run(combined_config)
 }
 
 /// Build the configuration needed to launch a test Runtime instance that runs a single-Node
 /// application with the given Wasm module file name and entrypoint.
```
Comment (rbehjati) [inline sdk/rust/oak_tests/src/lib.rs:135]: Update comment?

File: examples/abitest/oak_app_manifest.toml [hunk 6]
```diff
@@ -1,9 +1,9 @@
 name = "abitest"
 
 [modules]
-frontend_module = { path = "examples/abitest/bin/abitest_0_frontend.wasm" }
-backend_module = { path = "examples/abitest/bin/abitest_1_backend.wasm" }
-linear_handles_module = { path = "examples/abitest/bin/abitest_linear_handles.wasm" }
+frontend_module = { path = "examples/target/wasm32-unknown-unknown/release/abitest_0_frontend.wasm" }
+backend_module = { path = "examples/target/wasm32-unknown-unknown/release/abitest_1_backend.wasm" }
+linear_handles_module = { path = "examples/target/wasm32-unknown-unknown/release/abitest_linear_handles.wasm" }
```
Comment (anghelcovici) [inline examples/abitest/oak_app_manifest.toml:6]: This doesn't look nice at all. I clearly prefer the previous style and I am unsure why this change is needed here.

**Comment by anghelcovici** [comment]
`presumably because of some interference with externally published crates with the same names `
What does presumably mean here? Was this the reason that the tests were failing?

File: examples/abitest/oak_app_manifest.toml (hunk 6 above)
Comment (tiziano88) [reply examples/abitest/oak_app_manifest.toml:6]: This is the canonical location where the artifacts are built. The previous version was compiling everything twice , one for tests and one for the examples, without even reusing the intermediate cache. Basically test code and runner were doing the same job twice. The previous version was also not reusing cache between examples, because it ignored the workspace structure in which the examples are defined. Let me clarify in the commit message and in code comments.

Comment (anghelcovici) [reply examples/abitest/oak_app_manifest.toml:6]: How does the output path of the binary influences that? This feels strange

Comment (tiziano88) [reply examples/abitest/oak_app_manifest.toml:6]: It's the target dir that affects it, maybe we can get the best of both by making the bin dir relative to the workspace root, let me try that

File: sdk/rust/oak_tests/src/lib.rs (hunk 1 above)
Comment (tiziano88) [reply sdk/rust/oak_tests/src/lib.rs]: Done.

File: sdk/rust/oak_tests/src/lib.rs (hunk 4 above)
Comment (tiziano88) [reply sdk/rust/oak_tests/src/lib.rs]: Done.

File: sdk/rust/oak_tests/src/lib.rs (hunk 5 above)
Comment (tiziano88) [reply sdk/rust/oak_tests/src/lib.rs]: Done.

File: sdk/rust/oak_tests/src/lib.rs (hunk 3 above)
Comment (tiziano88) [reply sdk/rust/oak_tests/src/lib.rs:49]: Yes, I think all of them should be unified in #1965.

File: sdk/rust/oak_app_build/src/main.rs (hunk 2 above)
Comment (tiziano88) [reply sdk/rust/oak_app_build/src/main.rs]: Done.

File: examples/abitest/oak_app_manifest.toml (hunk 6 above)
Comment (tiziano88) [reply examples/abitest/oak_app_manifest.toml]: PTAL now, I think now it is a compromise between short paths and actually sharing caches in a meaningful way (though personally I still prefer the longer more explicit paths -- but let's have a separate discussion about that perhaps).

**Comment by tiziano88** [comment]
> `presumably because of some interference with externally published crates with the same names `
> What does presumably mean here? Was this the reason that the tests were failing?

Basically things started failing with weird errors about prost and oak_abi versions when I changed some dependencies, I didn't look into it closely, but in general we should always refer to local crates by paths rather than by version, so I did that in a few places. I will do a more thorough cleanup in a separate PR.

**Review by anghelcovici — APPROVED** [review]
